armi/reactor/compositeList.py

```diff
@@ -0,0 +1,246 @@
+
+    Many things in an ARMI :class:`~armi.reactor.reactors.Reactor`
+    can be seen as things stacked on things. An
+    :class:`~armi.reactor.assemblies.Assemblies` is a 1-D stack of
+    :class:`~armi.reactor.blocks.Block` instances, each with an axial
+    location and height. Fuel pins are another concept that would benefit
+    from a 1-D stack of :class:`~armi.reactor.composites.Composite` like things.
+
+    This class seeks to separate out storing and maintaing the axial mesh
+    of things (iteration, item access, mesh updating, etc.) from
+    assembly-specific tasks (moving assemblies within the reactor,
+    looking for coolant channels, etc.)
+
+    """
+
+    spatialGrid: Optional[Grid]
+
+    def __init__(self, name):
+        super().__init__(name)
+        self.spatialGrid = None
+
+    def __iter__(self) -> Iterator[CompositeWithHeight]:
+        """Iterate over all children"""
+        return super().__iter__()
+
+    def __getitem__(self, index: int) -> CompositeWithHeight:
+        """Return the child at a specific index"""
+        return super().__getitem__(index)
+
+    def add(self, obj: CompositeWithHeight):
+        """Add an item to the end of the list
+
+        .. note::
+
+            Object **must** at least have a ``getHeight`` method
+            to understand how the spatial grid needs to be updated
+
+        Parameters
+        ----------
+        obj : Composite
+            Object to be added to the end.
+
+        """
+        super().add(obj)
+        nItems = len(self)
+        # Assume spatial grid has been added already
+        obj.spatialLocator = self.spatialGrid[0, 0, nItems - 1]
+        zbounds = self.spatialGrid.getBounds()[2]
+        # Spatial grid has space for this component
+        if len(zbounds) < nItems:
+            zbounds[nItems] = zbounds[nItems - 1] + obj.getHeight()
+        else:
+            self.reestablishOrder()
+            self.calculateZCoords()
+
+    def insert(self, index: int, obj: CompositeWithHeight):
+        """Insert an object at a given position
+
+        Parameters
+        ----------
+        index : int
+            Position to insert the object
+        obj : CompositeWithHeight
+            Child to be inserted
+
+        """
+        super().insert(index, obj)
+        obj.spatialLocator = self.spatialGrid[0, 0, index]
+
+    def reestablishOrder(self):
+        """After children have been mixed up axially, place them in ascending order
+
+        See Also
+        --------
+        * :meth:`calculateZCoords` : Updates the zbottom / ztop parameters on each block
+
+        """
+        self.spatialGrid = axialUnitGrid(len(self), self)
+        for zi, child in enumerate(self):
+            child.spatialLocator = self.spatialGrid[0, 0, zi]
+            self._renameChild(child, zi)
+
+    @abstractmethod
+    def _renameChild(self, child: Composite, index: int):
+        """Rename a child at a given index in the list"""
+
+    def calculateZCoords(self):
```

Comment:
It would be cool if all these methods had their own unit tests.

Reply:
Excellent request. Since the `Assembly` is the primary concrete subclass, and it's well tested through `armi/reactor/tests/test_assemblies.py`, all if not most of these methods _should_ be tested.

The alternative approach could be to make another child subclass specific for testing the composite list methods. It would look similar to those in the assembly testing, but maybe not tied strictly to blocks? Ah but we need the children to have a `getHeight` method...

Thoughts?

Reply:
Ah, you're counting on the `Assembly` subclasses unit tests. I see.

Well, in that case, your [unit test coverage](https://coveralls.io/builds/55940866/source?filename=armi%2Freactor%2FcompositeList.py) is pretty high for the new file (~91%).

Reply:
If you're alright with it, I'll extend the existing assembly tests to bump the coverage up. There's some edge cases in some pretty public methods (`getChildrenAndZ` -> `getBlocksAndZ`) that aren't touched